### DISCUSSION TITLE
How to refresh tree-sitter parser after `vim.opt.rtp` has changed and new queries are now available?

### DISCUSSION THREAD

**Discussion by DanilaMihailov** [neovim/neovim#28708] [Q&A]
I'am adding tree-sitter queries to a plugin. Into directory `after/queries/lua/injections.scm`. Everything works great only if plugin is **not** loaded lazily.

I'am using lazy.nvim, but in any way it basically does `vim.opt.rtp:append("path/to/plugin")`. The problem is - before plugin loaded lazily, I already have tree-sitter parser running for buffer (filetype plugin already run and called `vim.treesitter.run()`). And tree-sitter reads query files once per "parser" (`vim.treesitter.LanguageTree`), but my plugin is not loaded yet, so `vim.opt.rtp` does not contain path to my plugin.

The only way I found to trigger reloading parser is by doing `:edit`, which drops parser for current buffer and creates new one. New parser now sees updated `runtimepath` and loads my queries.

I tried doing something like this

```lua
local parser = vim.treesitter.get_parser()
parser:destroy()
vim.treesitter.get_parser()
```

but `parser:destroy()` actually does nothing. And calls to `get_parser` return cached parser for current buffer. The only way parser is removed is by an event `nvim_buf_detach_event[{buf}]` (`:h nvim_buf_detach_event`). Which is triggered when buffer is closed or reloaded with `:edit`.

Is there a better way to refresh parser? Or another way to add queries via plugin that is **lazily** loaded?

**Comment by DanilaMihailov** [comment]
found a way to do this, but it requires use of private field

```lua
  local parser = vim.treesitter.get_parser()
  parser._injection_query = vim.treesitter.query.get("lua", "injections")
  parser:invalidate(true)
  parser:parse(true)
```

**Reply by okuuva** [reply]
I have the same issue, though I wasn't able to get it working even with setting `_injection_query`. The queries work fine if the plugin isn't lazy loaded, or if it's loaded before "nvim-treesitter/nvim-treesitter". Wouldn't want to make that a requirement for the plugin, the tree-sitter queries are a nice-to-have feature that ideally should Just Work™ if the user has tree-sitter set up.

Edit: the `:edit` workaround doesn't work for me either; haven't found any way to get the plugin provided queries to work if the plugin is loaded after "nvim-treesitter". Setting the queries with `vim.treesitter.query.set` would probably work, but it feels janky to read the queries from the files. It'd be much cleaner if queries (and parsers?) could be forcefully reloaded somehow.

**Reply by okuuva** [reply]
What is extra weird is that `vim.treesitter.query.get_files()` actually shows that the plugin provided query files are included even when the plugin is lazy loaded. But they still don't work for some reason...
